src/sentry/snuba/discover.py

```diff
@@ -770,82 +770,107 @@ def get_performance_facets(
 
-    # Exclude tracing tags as they are noisy and generally not helpful.
-    # TODO(markus): Tracing tags are no longer written but may still reside in DB.
-    excluded_tags = ["tags_key", "NOT IN", ["trace", "trace.ctx", "trace.span", "project"]]
-
-    # Sampling keys for multi-project results as we don't need accuracy
-    # with that much data.
-    sample = len(snuba_filter.filter_keys["project_id"]) > 2
-
     with sentry_sdk.start_span(op="discover.discover", description="facets.frequent_tags"):
         # Get the most relevant tag keys
         key_names = raw_query(
-            aggregations=[["count", None, "count"]],
+            aggregations=[
+                [aggregate_function, aggregate_column, "aggregate"],
+                ["count", None, "count"],
+            ],
             start=snuba_filter.start,
             end=snuba_filter.end,
             conditions=snuba_filter.conditions,
             filter_keys=snuba_filter.filter_keys,
-            orderby=["-count", "tags_key"],
-            groupby="tags_key",
-            # TODO(Kevan): Check using having vs where before mainlining
-            having=[excluded_tags],
+            orderby=["-count"],
             dataset=Dataset.Discover,
             limit=limit,
             referrer=referrer,
-            turbo=sample,
         )
-        top_tags = [r["tags_key"] for r in key_names["data"]]
-        if not top_tags:
+        counts = [r["count"] for r in key_names["data"]]
+        if not counts:
             return []
 
     results = []
     snuba_filter.conditions.append([aggregate_column, "IS NOT NULL", None])
 
-    # Only enable sampling if over 10000 values
-    sampling_enabled = key_names["data"][0]["count"] > 10000
-    options_sample_rate = options.get("discover2.tags_performance_facet_sample_rate") or 0.1
+    # Aggregate for transaction
+    transaction_aggregate = key_names["data"][0]["aggregate"]
 
-    sample_rate = options_sample_rate if sampling_enabled else None
+    # Dynamically sample so at least 10000 transactions are selected
+    transaction_count = key_names["data"][0]["count"]
+    sampling_enabled = transaction_count > 10000
+    target_sample = 10000 * (math.log(transaction_count, 10) - 3)  # Log growth starting at 10,000
 
-    max_aggregate_tags = 20
-    aggregate_tags = []
-    for i, tag in enumerate(top_tags):
-        if i >= len(top_tags) - max_aggregate_tags:
-            aggregate_tags.append(tag)
+    dynamic_sample_rate = 0 if transaction_count <= 0 else (target_sample / transaction_count)
+    options_sample_rate = (
+        options.get("discover2.tags_performance_facet_sample_rate") or dynamic_sample_rate
+    )
 
-    if orderby is None:
-        orderby = []
+    sample_rate = options_sample_rate if sampling_enabled else None
+    sample_multiplier = sample_rate if sample_rate else 1
+
+    excluded_tags = [
+        "tags_key",
+        "NOT IN",
+        ["trace", "trace.ctx", "trace.span", "project", "browser"],
```

Comment:
If we can filter out as many high cardinality keys as possible that'd help with the memory usage. This is because as the cardinality of the key approaches the total number of transactions, each value will only have a small sample size to work with, thus rendering the results not useful.

Reply:
Agree, lmk if you can think of any other high cardinality default tag keys we definitely don't need.

Reply:
- `url` is another high cardinality tag that won't tell us very much.
- `user` might be one? slowness at an individual user level might not be interesting enough to surface but I could be wrong
- `level` isn't high cardinality but I don't think that's very interesting to surface

Also tags with a cardinality of 1 also aren't helpful to show since they'll be on every transaction. Examples are
-  perhaps`routing.instrumentation`

Reply:
Thanks Tony, I'll copy these into my notes. I'm not sure about `user`, for example, since I think it could still be interesting or helpful when you have low transaction counts. I'm going to wait to check how it looks with prod data before removing more I think.